api/v1alpha6/conversion.go

```diff
@@ -17,24 +17,52 @@ limitations under the License.
 package v1alpha6
 
 import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	conversion "k8s.io/apimachinery/pkg/conversion"
 	utilconversion "sigs.k8s.io/cluster-api/util/conversion"
 	ctrlconversion "sigs.k8s.io/controller-runtime/pkg/conversion"
 
 	infrav1 "sigs.k8s.io/cluster-api-provider-openstack/api/v1alpha7"
 )
 
-var _ ctrlconversion.Convertible = &OpenStackCluster{}
+// Convert a source object of type S to dest type D using the provided conversion function.
+// Store the original source object in the dest object's annotations.
+// Also return any previous version of the object stored in the source object's annotations.
+func convertAndRestore[S, D metav1.Object](src S, dst D, previous D, convert func(S, D, conversion.Scope) error) (bool, error) {
+	// Restore data from previous conversion except for metadata.
+	// We do this before conversion because after convert() src.Annotations
+	// will be aliased as dst.Annotations and will therefore be modified by
+	// MarshalData below.
+	restored, err := utilconversion.UnmarshalData(src, previous)
+	if err != nil {
+		return false, err
+	}
 
-func (r *OpenStackCluster) ConvertTo(dstRaw ctrlconversion.Hub) error {
-	dst := dstRaw.(*infrav1.OpenStackCluster)
+	if err := convert(src, dst, nil); err != nil {
+		return false, err
+	}
 
-	if err := Convert_v1alpha6_OpenStackCluster_To_v1alpha7_OpenStackCluster(r, dst, nil); err != nil {
-		return err
+	// Store the original source object in the dest object's annotations.
+	if err := utilconversion.MarshalData(src, dst); err != nil {
+		return false, err
 	}
 
-	// Manually restore data.
-	restored := &infrav1.OpenStackCluster{}
-	if ok, err := utilconversion.UnmarshalData(r, restored); err != nil || !ok {
+	return restored, nil
+}
```

Comment:
@lentzi90 This is the change to the conversion functions I was talking about.

With this in place I don't think we will have to ignore any fields in the fuzzer tests because all fields should either be losslessly convertible or restored from the stashed value.

Reply:
Interesting! The first use of golang generics I have come across in the wild :smile: